File: typescript_test_crate/tests/13_system_return.ts
class SystemReturnTestFrameEvent {
    public _message: string;
    public _parameters: Record<string, any> | null;
    public _return: any;

    constructor(message: string, parameters: Record<string, any> | null) {
        this._message = message;
        this._parameters = parameters;
        this._return = null;
    }
}


class SystemReturnTestCompartment {
    public state: string;
    public state_args: Record<string, any>;
    public state_vars: Record<string, any>;
    public enter_args: Record<string, any>;
    public exit_args: Record<string, any>;
    public forward_event: any;
    public parent_compartment: SystemReturnTestCompartment | null;

    constructor(state: string, parent_compartment: SystemReturnTestCompartment | null = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    public copy(): SystemReturnTestCompartment {
        const c = new SystemReturnTestCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


class SystemReturnTest {
    private _state_stack: Array<any>;
    private __compartment: SystemReturnTestCompartment;
    private __next_compartment: SystemReturnTestCompartment | null;
    private _return_value: any;

    constructor() {
        this._state_stack = [];
        this._return_value = null;
        this.__compartment = new SystemReturnTestCompartment("Calculator");
        this.__next_compartment = null;
        const __frame_event = new SystemReturnTestFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    private __kernel(__e: SystemReturnTestFrameEvent) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new SystemReturnTestFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new SystemReturnTestFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new SystemReturnTestFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    private __router(__e: SystemReturnTestFrameEvent) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    private __transition(next_compartment: SystemReturnTestCompartment) {
        this.__next_compartment = next_compartment;
    }

    public add(a: number, b: number): number {
        this._return_value = null;
        const __e = new SystemReturnTestFrameEvent("add", {"0": a, "1": b});
        this.__kernel(__e);
        return this._return_value;
    }

    public multiply(a: number, b: number): number {
        this._return_value = null;
        const __e = new SystemReturnTestFrameEvent("multiply", {"0": a, "1": b});
        this.__kernel(__e);
        return this._return_value;
    }

    public greet(name: string): string {
        this._return_value = null;
        const __e = new SystemReturnTestFrameEvent("greet", {"0": name});
        this.__kernel(__e);
        return this._return_value;
    }

    public get_value(): number {
        this._return_value = null;
        const __e = new SystemReturnTestFrameEvent("get_value", null);
        this.__kernel(__e);
        return this._return_value;
    }

    private _state_Calculator(__e: SystemReturnTestFrameEvent) {
        if (__e._message === "$>") {
            this.__compartment.state_vars["value"] = 0;
        } else if (__e._message === "add") {
            const a = __e._parameters?.["0"];
            const b = __e._parameters?.["1"];
            this._return_value = a + b;
            __e._return = this._return_value;
            return;;
        } else if (__e._message === "get_value") {
            this.__compartment.state_vars["value"] = 42;
            this._return_value = this.__compartment.state_vars["value"];
            __e._return = this._return_value;
            return;;
        } else if (__e._message === "greet") {
            const name = __e._parameters?.["0"];
            this._return_value = "Hello, " + name + "!";
            __e._return = this._return_value;
            return;;
        } else if (__e._message === "multiply") {
            const a = __e._parameters?.["0"];
            const b = __e._parameters?.["1"];
            this._return_value = a * b;;
        }
    }
}


function main() {
    console.log("=== Test 13: System Return ===");
    const calc = new SystemReturnTest();

    // Test return sugar
    let result = calc.add(3, 5);
    if (result !== 8) {
        throw new Error(`Expected 8, got ${result}`);
    }
    console.log(`add(3, 5) = ${result}`);

    // Test system.return = expr
    result = calc.multiply(4, 7);
    if (result !== 28) {
        throw new Error(`Expected 28, got ${result}`);
    }
    console.log(`multiply(4, 7) = ${result}`);

    // Test string return
    const greeting = calc.greet("World");
    if (greeting !== "Hello, World!") {
        throw new Error(`Expected 'Hello, World!', got '${greeting}'`);
    }
    console.log(`greet('World') = ${greeting}`);

    // Test return with state variable
    const value = calc.get_value();
    if (value !== 42) {
        throw new Error(`Expected 42, got ${value}`);
    }
    console.log(`get_value() = ${value}`);

    console.log("PASS: System return works correctly");
}

main();

File: output/rust/tests/13_system_return.rs
use std::collections::HashMap;

#[derive(Clone, Debug)]
struct SystemReturnTestFrameEvent {
    message: String,
}

impl SystemReturnTestFrameEvent {
    fn new(message: &str) -> Self {
        Self { message: message.to_string() }
    }
}

#[derive(Clone)]
struct SystemReturnTestCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<SystemReturnTestFrameEvent>,
}

impl SystemReturnTestCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone, Default)]
struct CalculatorContext {
    value: i32,
}

#[derive(Clone)]
enum SystemReturnTestStateContext {
    Calculator(CalculatorContext),
    Empty,
}

impl Default for SystemReturnTestStateContext {
    fn default() -> Self {
        SystemReturnTestStateContext::Calculator(CalculatorContext::default())
    }
}

pub struct SystemReturnTest {
    _state_stack: Vec<(String, SystemReturnTestStateContext)>,
    __compartment: SystemReturnTestCompartment,
    __next_compartment: Option<SystemReturnTestCompartment>,
    _sv_value: i32,
}

impl SystemReturnTest {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            _sv_value: 0,
            __compartment: SystemReturnTestCompartment::new("Calculator"),
            __next_compartment: None,
        };
let __frame_event = SystemReturnTestFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: SystemReturnTestFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = SystemReturnTestFrameEvent::new("$<");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = SystemReturnTestFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = SystemReturnTestFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &SystemReturnTestFrameEvent) {
match self.__compartment.state.as_str() {
    "Calculator" => self._state_Calculator(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: SystemReturnTestCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Calculator" => SystemReturnTestStateContext::Calculator(CalculatorContext { value: self._sv_value }),
    _ => SystemReturnTestStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = SystemReturnTestFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    SystemReturnTestStateContext::Calculator(ctx) => {
        self._sv_value = ctx.value;
    }
    SystemReturnTestStateContext::Empty => {}
}
    }

    pub fn add(&mut self, a: i32, b: i32) -> i32 {
let __e = SystemReturnTestFrameEvent::new("add");
match self.__compartment.state.as_str() {
            "Calculator" => self._s_Calculator_add(&__e, a, b),
            _ => Default::default(),
        }
    }

    pub fn multiply(&mut self, a: i32, b: i32) -> i32 {
let __e = SystemReturnTestFrameEvent::new("multiply");
match self.__compartment.state.as_str() {
            "Calculator" => self._s_Calculator_multiply(&__e, a, b),
            _ => Default::default(),
        }
    }

    pub fn get_value(&mut self) -> i32 {
let __e = SystemReturnTestFrameEvent::new("get_value");
match self.__compartment.state.as_str() {
            "Calculator" => self._s_Calculator_get_value(&__e),
            _ => Default::default(),
        }
    }

    fn _state_Calculator(&mut self, __e: &SystemReturnTestFrameEvent) {
match __e.message.as_str() {
    "get_value" => { self._s_Calculator_get_value(__e); }
    "$>" => {
        self._sv_value = 0;
    }
    _ => {}
}
    }

    fn _s_Calculator_add(&mut self, __e: &SystemReturnTestFrameEvent, a: i32, b: i32) -> i32 {
return a + b
    }

    fn _s_Calculator_get_value(&mut self, __e: &SystemReturnTestFrameEvent) -> i32 {
self._sv_value = 42;
return self._sv_value
    }

    fn _s_Calculator_multiply(&mut self, __e: &SystemReturnTestFrameEvent, a: i32, b: i32) -> i32 {
return a * b
    }
}


fn main() {
    println!("=== Test 13: System Return ===");
    let mut calc = SystemReturnTest::new();

    // Test caret return sugar
    let result = calc.add(3, 5);
    assert_eq!(result, 8, "Expected 8, got {}", result);
    println!("add(3, 5) = {}", result);

    // Test system.return = expr
    let result = calc.multiply(4, 7);
    assert_eq!(result, 28, "Expected 28, got {}", result);
    println!("multiply(4, 7) = {}", result);

    // Test return with state variable
    let value = calc.get_value();
    assert_eq!(value, 42, "Expected 42, got {}", value);
    println!("get_value() = {}", value);

    println!("PASS: System return works correctly");
}
